[LoopVectorize][LAA] Hoist load in memory IV to allow vectorization

felipealmeida opened the pull request:
# Goal

Adds VPScalarIVPromotionRecipe recipe to promote memory IV to scalar
IV.

Use SCEV mutilplied by VFxUF or EVL to multiply the SCEV between load
and store.

This PR addresses issue #165312

# Example

The code that this patch allows vectorization is like this:

```
while.body:
  %theFirst.addr.0112 = phi ptr [ %incdec.ptr9, %while.body ], [ %theFirst, %while.body.preheader ]
  %thePointer.0111 = phi ptr [ %incdec.ptr, %while.body ], [ %add.ptr.i, %while.body.preheader ]
  %1 = load i16, ptr %theFirst.addr.0112, align 2
  store i16 %1, ptr %thePointer.0111, align 2
  %incdec.ptr = getelementptr inbounds nuw i8, ptr %thePointer.0111, i64 2
  %2 = load i64, ptr %m_size_ptr, align 8
  %inc = add i64 %2, 1
  store i64 %inc, ptr %m_size_ptr, align 8
  %incdec.ptr9 = getelementptr inbounds nuw i8, ptr %theFirst.addr.0112, i64 2
  %cmp7.not = icmp eq ptr %incdec.ptr9, %theLast
  br i1 %cmp7.not, label %cleanup.loopexit, label %while.body
```

As you can see, %m_size_ptr is a loop invariant pointer and can be
promoted to a IV scalar and then execute vectorization.

# C++ code

The C++ code from Xalan that generates the LLVM IR above is this:

```
        const size_type     theInsertSize =
            local_distance(theFirst, theLast);
        if (theInsertSize == 0)
        {
            return;
        }
        const size_type     theTotalSize = size() + theInsertSize;
        if (thePosition == end())
        {
            pointer     thePointer = ensureCapacity(theTotalSize);
            while (theFirst != theLast)
            {
                Constructor::construct(thePointer, *theFirst, *m_memoryManager);
                ++thePointer;
                ++m_size;
                ++theFirst;
            }
        }
```

# Results

On SPEC CPU 2017 for RISC-V on Ventana design we can see a 25% improvement for Xalan benchmark.

On SPEC CPU 2017 for AArch64 on the Raspberry pi 5 we can see a 17% improvement for Xalan benchmark.

fhahn reviewed (commented):
I'm curious what the original C/C++ source code looks like. Is there anything we are missing to enable hoisting/sinking earlier?

After a first look, I am not sure LAA is the best place to add this hoisting detection, as it adds quite a bit of complexity and coupling spread out across LAA and LV. But I need to think about it a bit more to see what possible alternatives are.

fhahn commented:
> Hello Florian, thanks for taking the time to review it.
> 
> > I'm curious what the original C/C++ source code looks like. Is there anything we are missing to enable hoisting/sinking earlier?
> > After a first look, I am not sure LAA is the best place to add this hoisting detection, as it adds quite a bit of complexity and coupling spread out across LAA and LV. But I need to think about it a bit more to see what possible alternatives are.
> 
> I added the C++ code to the PR message. I don't think this hoisting can be made before LoopVectorize. I did look into that. However, it is not possible to prove that m_size pointer doesn't alias with the thePointer and theFirst. It is possible to hoist in the LoopVectorize because it can add a runtime check.
> 

Would it be possible to share a build-able reproducer? From looking at the code, I am surprised that we are not able to determine no-alias via TBAA metadata given that it looks like we are accessing 16 and 64 bit integers, which should be have different TBAA nodes.

fhahn commented on llvm/test/Transforms/LoopVectorize/memory-iv-promotion.ll:
```diff
@@ -0,0 +1,60 @@
+; RUN: opt < %s -passes=loop-vectorize -mtriple=x86_64-unknown-linux-gnu -S | FileCheck %s --check-prefix=X86_64
+
+; Testcase extraído de ElemAttribute.cpp
+; Foca no loop while.body que copia elementos i16
```
comment needs cleaning up

fhahn reviewed (commented):
Thanks for the update. it looks like there are a number of  unrelated/uneeded changes which woud be good ot remove to reduce the diff.

I am wondering if it would make sense to implement this as separate transform, as it seems like this may be useful also in some cases we don't vectorize? We already have LoopVersioningLICM which does something similar, although I don't think it would hoist the load + sink the store yet.

felipealmeida committed:
[LoopVectorize] Add baseline test for memory induction variable promotion

Add test cases for loops containing a load/add/store pattern to a
loop-invariant address (memory induction variable) alongside array
operations. Currently these loops are not vectorized due to the
load/store dependence on the invariant address.

Test cases:
- memory_iv_copy_and_count: array copy with ++counter (step 1)
- memory_iv_step3: array copy with counter += 3
- memory_iv_external_use: negative test where counter value has
  external users (call), preventing promotion

felipealmeida committed:
[LoopVectorize] Promote memory induction variables to scalar IVs

Enable vectorization of loops with memory induction variables —
load/step/store patterns to loop-invariant addresses (e.g., ++m_size)
that previously caused vectorization to be rejected due to load/store
dependences involving invariant addresses.

Detection (LoopVectorizationLegality):
- Add isInvariantLoadHoistable() which uses MemorySSA to verify no
  other clobbers exist and SCEV to compute the loop-invariant step
- Detect memory IVs in canVectorizeMemory() and allow vectorization
  when all invariant-address conflicts are promotable

VPlan transform (VPlanTransforms::promoteMemoryIVs):
For each detected memory IV:
1. Hoist the load to the preheader (single scalar VPReplicateRecipe)
2. Create a scalar PHI in the loop header
3. Compute the latch update: PHI + Step * VFxUF
4. Store the final accumulated value in the middle block
5. Remove the original load/store recipes; dead intermediates are
   cleaned up by removeDeadRecipes

The transform runs after addExplicitVectorLength/optimizeEVLMasks to
avoid conflicting with VFxUF user assertions.

Also adds Instruction::Store support in VPInstruction (generate,
isSingleScalar, usesFirstLaneOnly, usesFirstPartOnly) and
Instruction::PHI to usesFirstPartOnly for correct unroller behavior.

felipealmeida commented:
> Thanks for the update. it looks like there are a number of unrelated/uneeded changes which woud be good ot remove to reduce the diff.
> 
> I am wondering if it would make sense to implement this as separate transform, as it seems like this may be useful also in some cases we don't vectorize? We already have LoopVersioningLICM which does something similar, although I don't think it would hoist the load + sink the store yet.

I think this version fixes all or most of your objections
